pkg/gather/ssh/ssh.go

```diff
@@ -141,3 +191,84 @@ func LoadPrivateSSHKeys(paths []string) (map[string]interface{}, error) {
 	}
 	return keys, nil
 }
+
+// newHTTPProxy creates a new httpProxy populated with credentials
+// if any is available
+func newHTTPProxy(uri *url.URL, forward proxy.Dialer) (proxy.Dialer, error) {
+	proxyDialer := new(httpProxy)
+	proxyDialer.host = uri.Host
+	proxyDialer.forward = forward
+	if uri.User != nil {
+		proxyDialer.haveAuth = true
+		proxyDialer.username = uri.User.Username()
+		proxyDialer.password, _ = uri.User.Password()
+	}
+
+	return proxyDialer, nil
+}
+
+// newHTTPSProxy creates a new httpProxy populated with credentials using HTTPS dialer
+func newHTTPSProxy(uri *url.URL, forward proxy.Dialer) (proxy.Dialer, error) {
+	return newHTTPProxy(uri, HttpsDialer)
+}
+
+// Dial is a custom dialer for the ssh client that will use HTTP CONNECT method via TLS handshake
+func (d httpsDialer) Dial(network, addr string) (c net.Conn, err error) {
+	c, err = tls.Dial("tcp", addr, TlsConfig)
+	if err != nil {
+		fmt.Println(err)
+	}
+	return
+}
+
+// Dial is a custom dialer for the ssh client that will use HTTP CONNECT method
+// to establish a connection to the proxy. This is needed because the net/proxy module
+// only supports SOCKS5 proxies.
+func (s *httpProxy) Dial(network, addr string) (net.Conn, error) {
+	c, err := s.forward.Dial("tcp", s.host)
+	if c == nil {
+		return nil, errors.Wrap(err, "failed to connect to proxy")
+	}
+	if err != nil {
+		c.Close()
+		return nil, err
+	}
+
+	reqURL, err := url.Parse("http://" + addr)
+	if err != nil {
+		c.Close()
+		return nil, err
+	}
+	reqURL.Scheme = ""
+
+	req, err := http.NewRequest("CONNECT", reqURL.String(), nil)
+	if err != nil {
+		c.Close()
+		return nil, err
+	}
+	req.Close = false
+	if s.haveAuth {
+		req.SetBasicAuth(s.username, s.password)
+	}
+
+	err = req.Write(c)
+	if err != nil {
+		c.Close()
+		return nil, err
+	}
+
+	resp, err := http.ReadResponse(bufio.NewReader(c), req)
+	if err != nil {
+		c.Close()
+		return nil, err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != 200 {
+		logrus.Debugf("connect server using proxy, StatusCode [%d]", resp.StatusCode)
```

Comment:
since we're returning the error, maybe there is no need for the log?

Reply:
right. i'll put the status code however.